### PR TITLE
Newly add test cases to cover new feature support in virsh snapshot

### PR DESCRIPTION
**Pull request by chunfuwen**
New snapshot mode: mode="manual" is introduced since libvirt version (8, 2, 0)

Signed-off-by: chunfuwen <chwen@redhat.com>

### CODE REVIEW

**Comment by chunfuwen** [comment]
PASS 1-type_specific.io-github-autotest-libvirt.virsh.snapshot_mode.coldplug.external_manual_mode.positive_test
PASS 2-type_specific.io-github-autotest-libvirt.virsh.snapshot_mode.coldplug.external_manual_mode_memdump.positive_test
PASS 3-type_specific.io-github-autotest-libvirt.virsh.snapshot_mode.coldplug.external_manual_mode_with_specified_memory_file.positive_test

**Comment by chunfuwen** [comment]
@lento-sun please help take a look

File: libvirt/tests/src/virsh_cmd/snapshot/virsh_snapshot_mode.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+        if snapshot_name in value:
+            cleanup_files.append(value)
+        if item in ['vdc'] and snapshot_name in value:
+            test.fail("Wrongly create snapshot on vdc")
+    # Validate from snapshot dumpxml
+    for key, value in dump_snapshot_xml(vm.name, snapshot_name).items():
+        if key in ['vdb'] and value not in ['external']:
+            test.fail("snapshot attribute is wrong from vdb")
+        if key in ['vdc'] and value not in ['manual']:
+            test.fail("snapshot attribute is wrong from vdc")
```
Comment (dzhengfy) [inline libvirt/tests/src/virsh_cmd/snapshot/virsh_snapshot_mode.py:171]: same as above

File: libvirt/tests/src/virsh_cmd/snapshot/virsh_snapshot_mode.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+                vmxml.add_device(device_obj)
+                vmxml.sync()
+            else:
+                if not vm.is_alive():
+                    vm.start()
+                    vm.wait_for_login().close()
+                LOG.info("Attaching devices...")
+                result = virsh.attach_device(vm_name, device_obj.xml, debug=True)
+                libvirt.check_result(result, expected_fails=[])
+        # Start VM when cold plug
```
Comment (dzhengfy) [inline libvirt/tests/src/virsh_cmd/snapshot/virsh_snapshot_mode.py:227]: Here let me confuse. Shouldn't we add the checking if it is coldplug?

**Review by dzhengfy — COMMENTED** [review]
Others LGTM

**Comment by chunfuwen** [comment]
@lento-sun 
This is related to add 'manual' mode feature in snapshot, please help have a review

**Comment by lento-sun** [comment]
> @lento-sun This is related to add 'manual' mode feature in snapshot, please help have a review

working on this, thx

File: libvirt/tests/src/virsh_cmd/snapshot/virsh_snapshot_mode.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    for disk in disks:
+        if disk.device != 'disk':
+            disks.remove(disk)
+    new_disks = []
+    for src_disk_xml in disks:
+        disk_xml = snap_xml.SnapDiskXML()
+        disk_target = src_disk_xml.target['dev']
+        disk_xml.disk_name = disk_target
+        if 'vda' in disk_target:
+            #disk_xml.name = "vda"
```
[the remark this answers is not in the document]
Comment (chunfuwen) [reply libvirt/tests/src/virsh_cmd/snapshot/virsh_snapshot_mode.py:108]: Done

File: libvirt/tests/src/virsh_cmd/snapshot/virsh_snapshot_mode.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+            disk_xml.snapshot = "external"
+        if 'vdb' in disk_target:
+            #disk_xml.name = "vdb"
```
[the remark this answers is not in the document]
Comment (chunfuwen) [reply libvirt/tests/src/virsh_cmd/snapshot/virsh_snapshot_mode.py:111]: Done

File: libvirt/tests/src/virsh_cmd/snapshot/virsh_snapshot_mode.py [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+            disk_xml.snapshot = "external"
+        elif 'vdc' in disk_target:
+            #disk_xml.name = "vdc"
```
[the remark this answers is not in the document]
Comment (chunfuwen) [reply libvirt/tests/src/virsh_cmd/snapshot/virsh_snapshot_mode.py:114]: Done

File: libvirt/tests/src/virsh_cmd/snapshot/virsh_snapshot_mode.py [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+    dom_blk_dict = translate_domblklist_output_into_dict(vm.name)
+    for item, value in dom_blk_dict.items():
+        if snapshot_name in value:
+            cleanup_files.append(value)
+        if item in ['vdc'] and snapshot_name in value:
+            test.fail("Wrongly create snapshot on vdc")
+    # Validate from snapshot dumpxml
+    for key, value in dump_snapshot_xml(vm.name, snapshot_name).items():
+        if key in ['vdb'] and value not in ['external']:
+            test.fail("snapshot attribute is wrong from vdb")
```
[the remark this answers is not in the document]
Comment (chunfuwen) [reply libvirt/tests/src/virsh_cmd/snapshot/virsh_snapshot_mode.py:169]: updated as mentioned

File: libvirt/tests/src/virsh_cmd/snapshot/virsh_snapshot_mode.py (hunk 1 above)
Comment (chunfuwen) [reply libvirt/tests/src/virsh_cmd/snapshot/virsh_snapshot_mode.py:171]: updated as mentioned

File: libvirt/tests/src/virsh_cmd/snapshot/virsh_snapshot_mode.py (hunk 2 above)
Comment (chunfuwen) [reply libvirt/tests/src/virsh_cmd/snapshot/virsh_snapshot_mode.py:227]: 1) Do you mean checking disk after cold plug? logic is put in previous code(around line 214):
```
            if not hotplug:
                vmxml.add_device(device_obj)
                vmxml.sync()
```
2)Here it is normal logic that VM is not started after cold plug disk, so we need start it to meet subsequent operations

**Review by dzhengfy — COMMENTED** [review]
LGTM from codes perspective.

**Review by dzhengfy — APPROVED** [review]

File: libvirt/tests/src/virsh_cmd/snapshot/virsh_snapshot_mode.py [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+    # Remove non-storage disk such as 'cdrom'
+    for disk in disks:
+        if disk.device != 'disk':
+            disks.remove(disk)
+    new_disks = []
+    for src_disk_xml in disks:
+        disk_xml = snap_xml.SnapDiskXML()
+        disk_target = src_disk_xml.target['dev']
+        disk_xml.disk_name = disk_target
+        if 'vda' in disk_target:
```
[the remark this answers is not in the document]
Comment (chunfuwen) [reply libvirt/tests/src/virsh_cmd/snapshot/virsh_snapshot_mode.py:107]: @lento-sun 
Thank you for suggestion.
But in this situation, disk_xml.snapshot will be initialized under already existed loop in line 103.
```
for src_disk_xml in disks:
    xxx
     disk_xml.snapshot
```  
It seems that your suggestion above is independent loop.

File: libvirt/tests/src/virsh_cmd/snapshot/virsh_snapshot_mode.py [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+    # Validate from virsh domblklist command
+    dom_blk_dict = translate_domblklist_output_into_dict(vm.name)
+    for item, value in dom_blk_dict.items():
+        if snapshot_name in value:
+            cleanup_files.append(value)
+        if item in ['vdc'] and snapshot_name in value:
+            test.fail("Wrongly create snapshot on vdc")
+    # Validate from snapshot dumpxml
+    for key, value in dump_snapshot_xml(vm.name, snapshot_name).items():
+        if key in ['vdb'] and value not in ['external']:
```
[the remark this answers is not in the document]
Comment (chunfuwen) [reply libvirt/tests/src/virsh_cmd/snapshot/virsh_snapshot_mode.py:165]: @lento-sun 
Thank you for suggestion.
Here I use dump_snapshot_xml to dynamically get snapshots,which are validated steps in polarion manual cases

File: libvirt/tests/cfg/virsh_cmd/snapshot/virsh_snapshot_mode.cfg [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    take_regular_screendumps = "no"
+    start_vm = "no"
+    target_bus = "virtio"
+    target_format = "qcow2"
+    type_name = "file"
+    driver_type = 'qcow2'
+    device_type = "disk"
+    func_supported_since_libvirt_ver = (8, 2, 0)
+    virt_disk_device_source = "/var/lib/libvirt/images/test.img"
+    snapshot_file = "/tmp/vm1.mem"
```
[the remark this answers is not in the document]
Comment (chunfuwen) [reply libvirt/tests/cfg/virsh_cmd/snapshot/virsh_snapshot_mode.cfg:12]: updated as suggested

File: libvirt/tests/src/virsh_cmd/snapshot/virsh_snapshot_mode.py [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+            LOG.info("As expected, failed to define VM due to reason:\n%s", str(xml_error))
+    except Exception as ex:
+        test.fail("unexpected exception happen: %s" % str(ex))
+    else:
+        if backend_device in ["external_manual_mode"]:
+            check_vm_paused(params, vm, test)
+        elif backend_device in ["external_manual_mode_memdump"]:
+            snapshot_name = params.get('snapshot_name')
+            snapshot_file = params.get('snapshot_file')
+            snapshot_opt = ("%s --diskspec vda,snapshot=external --diskspec vdb,snapshot=external"
```
[the remark this answers is not in the document]
Comment (chunfuwen) [reply libvirt/tests/src/virsh_cmd/snapshot/virsh_snapshot_mode.py:244]: updated almost like suggested one

**Comment by chunfuwen** [comment]
After the change, new test result:
Command 'virsh list' finished with 0 after 0.031685958s
cleaning libvirtd logs...
DATA (filename=output.expected) => NOT FOUND (data sources: variant, test, file)
DATA (filename=stdout.expected) => NOT FOUND (data sources: variant, test, file)
DATA (filename=stderr.expected) => NOT FOUND (data sources: variant, test, file)
PASS 1-type_specific.io-github-autotest-libvirt.virsh.snapshot_mode.coldplug.external_manual_mode_memdump.positive_test

**Comment by chunfuwen** [comment]
@lento-sun 
Comments are updated, please double check this.

**Comment by lento-sun** [comment]
lgtm, the script covered cases scenarios

**Review by lento-sun — APPROVED** [review]

**Pull request merged**